### ISSUE TITLE
Consider increasing trigger_data bit amount

### ISSUE THREAD

**Issue by arielhal** [WICG/attribution-reporting-api#733]
Hello,
Currently trigger_data size is limited to 3 bits (8 distinct values) when registering a trigger for event-level reports.
We would like to know what is the reason for such a strict limitation? and if it's possible to add few more bits to this value?
Many of our customers have more than 8 conversion types, and such limitation will prevent us from giving our customers a precise evaluation of their performance.

Thanks!

**Comment by csharrison** [comment]
Hello @arielhal , the reason the size is limited to 8 distinct values is to ensure that the amount of cross-site/context information about a user is limited. This is especially true for event-level reports, because they come associated with a high entropy identifier that links a report to the specific ad engagement.

I have two suggestions:
1. Look into the API's support for [aggregatable reports](https://github.com/WICG/attribution-reporting-api/blob/main/AGGREGATE.md). With these, you can make different tradeoffs about the amount of "conversion-side" information embedded in a report.
2. Consider taking a look and providing feedback on this issue: https://github.com/WICG/attribution-reporting-api/issues/694. In particular, I've linked a calculator into that issue which will also explain the amount of noise needed to satisfy (one of the) privacy bars we have in the API, based on varying parameters like the trigger_data cardinality.

**Comment by csharrison** [comment]
Hello @arielhal I also want to bring your attention to a document we recently published that allows you to trade off bits with other parameters in the event-level reports. Please take a look:
https://github.com/WICG/attribution-reporting-api/blob/main/flexible_event_config.md

I believe this should satisfy the feature request if we move forward with the proposal, can you confirm?

**Comment by arielhal** [comment]
Hey @csharrison !
Sorry for the late response, it looks good!
Can we define a trigger_data to be in the range of [0, .., 10] for example instead of the default [0,...,7]
What will be the impact? Where will other bits be reduced? Or will it just add more noise?
Also, where do we send the trigger_spec? in each trigger event to the browser?
Is there any estimation for when this feature will be available?

**Comment by apasel422** [comment]
Closing as this is being addressed with the Full Flex proposal.

**Comment by arielhal** [comment]
Hey @csharrison !
Is there an estimation for when Phase 2 (Full Flexible Event-Level) will be available?

Thanks!

**Comment by csharrison** [comment]
@arielhal we don't have a concrete timeline yet, but will update as the feature gets closer. We are in the process of specifying it right now.

**Comment by arielhal** [comment]
Hey @csharrison how are you? :)
Checking again about the ETA for Phase 2 (Full Flexible Event-Level)
Do you have any estimation?
Thanks!

**Comment by csharrison** [comment]
Thanks for the ping. We don't have a concrete timeline yet and are still in the process of designing and specifying the right behavior.

**Comment by pmainguet** [comment]
Hello @csharrison we are also interested to use the Phase 2 of Full Flexible Event-Level.
Any progress on that one since Jan 3rd ?

**Comment by csharrison** [comment]
Hey @pmainguet is your interest in Phase 2 for the use-case described in this issue or something else?

**Comment by apasel422** [comment]
Chromium expects [custom trigger data](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#data-limits-and-noise) (values and cardinality) to be available in M123.

**Comment by omriariav** [comment]
We appreciate the work and we're excited to use the custom trigger data (full flexible event-level).

**Comment by pmainguet** [comment]
@csharrison we work in affiliate marketing where users clicks on links and are then redirected to merchant pages to make salesvia a redirect through our server to count clicks (without user interaction).

Currently, we are very limited by the amount of information we can pass to the API on the click/source side (1 bit of data in our case).

Moreover, to calculate the performance of our different source of traffic, we would need to somehow cross summary reports with event level reports to be able to compute a sales/clicks aggregate, which is highly impractical and will be awfully noised.

The fully flexible event-level feature is exciting for us as we would have more bits of data available and we won't need to use summary reports to do performance optimisation.

And like @omriariav from Taboola we would ideally also need additional info such as the orderid so that our partners trust our computation and reporting.

**Comment by danieldansker** [comment]
Hey @csharrison  how are you?
Regarding the usage of Flexible Event-Level, with the parameter **"trigger_data_matching": "exact".**
Is it correct that in order to have a match we need to add to the source-event a list of integers that must have an exact match during the trigger-event.
Example:

**Attribution-Reporting-Register-Source:**
```
   {
        "source_event_id":"0000000",
        "destination":"https://test.com",
        "trigger_data_matching":"exact",
        "trigger_specs":[{
            "trigger_data":[111111,222,333]
        }]
}
```

**Attribution-Reporting-Register-Trigger:**
```
{
  "event_trigger_data": [
    {
      "trigger_data": 111111
    }
  ]
}
```

I am testing this on beta browser  - **Version 123.0.6312.22 (Official Build) beta (arm64)**, but all reports that are sent are still with a 3-bit limitation.
this is the source that is saved in attribution-internals: (as you can see the trigger data is limited)
```
{
   "aggregatable_report_window": 2592000,
   "debug_reporting": false,
   "destination": "https://round-table.co.il",
   "event_level_epsilon": 14.0,
   "expiry": 2592000,
   "max_event_level_reports": 3,
   "priority": "0",
   "source_event_id": "7172822603428505672",
   "trigger_data_matching": "exact",
   "trigger_specs": [ {
      "event_report_windows": {
         "end_times": [ 172800, 604800, 2592000 ],
         "start_time": 0
      },
      "trigger_data": [ 0, 1, 2, 3, 4, 5, 6, 7 ]
   } ]
}
```

**Comment by apasel422** [comment]
@danieldansker In your source registration, this is incorrect:

```jsonc
"trigger_specs":[{
  "trigger_data":[111111,222,333]
}]
```

Chromium doesn't yet support top-level `trigger_specs`: You should move the `trigger_data` field to the top-level of the source registration, i.e. as a sibling of `trigger_data_matching` [per the explainer](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#data-limits-and-noise).

**Comment by apasel422** [comment]
[Here is a header-validator link](https://wicg.github.io/attribution-reporting-api/validate-headers?header=source&json=++%7B%0A++++++++%22source_event_id%22%3A%220000000%22%2C%0A++++++++%22destination%22%3A%22https%3A%2F%2Ftest.com%22%2C%0A++++++++%22trigger_data_matching%22%3A%22exact%22%2C%0A++++++++%22trigger_specs%22%3A%5B%7B%0A++++++++++++%22trigger_data%22%3A%5B111111%2C222%2C333%5D%0A++++++++%7D%5D%0A%7D&source-type=event&flex=false) warning about the unknown `trigger_specs` field.

**Comment by omriariav** [comment]
Hi - we see that orderId is now supported in the 'flex event' (2nd bullet under 'goals' [here](https://github.com/WICG/attribution-reporting-api/blob/main/flexible_event_config.md#goals)) 

We are going to test this in the coming weeks.

Do you happen to have any updates regarding the currency field? 

(cc: @danieldansker )

**Comment by apasel422** [comment]
@omriariav Could you clarify what you mean by orderId and the currency field? Perhaps you mean the value field mentioned in the proposal? If so, it hasn’t been specified yet.

**Comment by taboola1dsg** [comment]
@apasel422 you can see clarification in Omri's [previous reply ](https://github.com/WICG/attribution-reporting-api/issues/733#issuecomment-1964452926).

When are you planning to add support to these parameters/values? thanks

**Comment by danieldansker** [comment]
Hey @apasel422 ,
1) I am trying to understand the exact noise mechanism for event-level reports and can't really find information.
2) Understand the triad-off between trigger-data list-size and max_event_level_reports. How can I check and test this?
Thanks

**Comment by apasel422** [comment]
@danieldansker

> I am trying to understand the exact noise mechanism for event-level reports and can't really find information.

An overview is available in [the explainer](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#data-limits-and-noise). More details are available in [the specification](https://wicg.github.io/attribution-reporting-api/#obtain-a-randomized-source-response).

> Understand the trade-off between trigger-data list-size and max_event_level_reports. How can I check and test this?

The simplest option is probably [the header validator](https://wicg.github.io/attribution-reporting-api/validate-headers?header=source&json=%7B%0A++%22destination%22%3A+%22https%3A%2F%2Fa.example%22%2C%0A++%22trigger_data%22%3A+%5B0%2C+1%2C+2%5D%2C%0A++%22max_event_level_reports%22%3A+3%0A%7D&source-type=event&flex=false).

**Comment by danieldansker** [comment]
Hi @apasel422 ,
Can we change the 'event_level_epsilon' parameter value?
In addition can you please explain how the informatio_gain is calculated?
I couldn't find the documentation.
Thanks

**Comment by apasel422** [comment]
> Can we change the 'event_level_epsilon' parameter value?

Yes. [Here's an example](https://wicg.github.io/attribution-reporting-api/validate-headers?header=source&json=%7B%22destination%22%3A%22https%3A%2F%2Fx.example%22%2C%22event_level_epsilon%22%3A5%7D&source-type=event&flex=false).

> In addition can you please explain how the informatio_gain is calculated?

[Here is the specification definition](https://wicg.github.io/attribution-reporting-api/#compute-the-channel-capacity-of-a-source).

You may find it easier, though, to look at an actual implementation: [TypeScript](https://github.com/WICG/attribution-reporting-api/blob/f41f1955530d1ae7ecfb5cc90deb06fdbda982d6/ts/src/flexible-event/privacy.ts#L151), [C++](https://source.chromium.org/chromium/chromium/src/+/main:components/attribution_reporting/privacy_math.cc;l=540;drc=9f557da02029d998ea8dc224b6681bedcbd3ad18).
